lightwood/api/predictor.py

```diff
@@ -136,13 +136,71 @@ def predict(self, data: pd.DataFrame, args: Dict[str, object] = {}) -> pd.DataFr
         """  # noqa
         pass
 
-    def save(self, file_path: str) -> None:
+
+
+        """
+            NOTE FROM CONTRIBUTOR - will remove before finalizing
+
+            I spotted this code snippet in the custom_cleaner.ipynb file
+            I figured we could maybe have the user pass it optionally
+            in instead of calling code_to_json_ai so it doesn't break
+            current expected functionality. Let me know what you think,
+            and sorry if I'm missing something obvious. I'm still
+            working on figuring out the codebase I only started
+            looking into the repo a few days ago.
+
+            code = code_from_json_ai(json_ai)
+            # Save code to a file (Optional)
+            with open('custom_cleaner_pipeline.py', 'w') as fp:
+            fp.write(code)
+
+            Here was my initial starting point for outlining the idea:
+
+            import json
+            import dill
+
+            class Predictor:
+                # Existing code ...
+
+                def save(self, file_path: str) -> None:
+
+                    # With a provided file path, saves the Predictor instance for later use.
+                    # :param file_path: Location to store your Predictor Instance.
+                    # :returns: Saves Predictor instance.
+
+                    predictor_dict = {}
+                    predictor_dict['predictor'] = self.__dict__
+                    predictor_dict['code'] = json.loads(self.to_json())
+
+                    with open(file_path, "wb") as fp:
+                        dill.dump(predictor_dict, fp)
+
+
+            Original code for save method:
+            def save(self, file_path: str) -> None:
+                # With a provided file path, saves the Predictor instance for later use.
+                # :param file_path: Location to store your Predictor Instance.
+                # :returns: Saves Predictor instance.
+                with open(file_path, "wb") as fp:
+                    dill.dump(self, fp)
+
+        """
+
+
+    def save(self, file_path: str, json_ai_code: Optional[str] = None) -> None:
         """
         With a provided file path, saves the Predictor instance for later use.
 
         :param file_path: Location to store your Predictor Instance.
+        :param code: The code generated by the user's specification (optional)
 
         :returns: Saves Predictor instance.
         """
+        predictor_dict = {}
```

Comment:
This idea is overall on the right track.

I think the one missing bit is to actually produce the json_ai internally, otherwise there is a risk in having a mismatch between the passed json_ai and what `self` actually looks like. This is a change that requires a bit more tinkering though, so let's leave it for a second PR.

Reply:
you got it! I'll get this done soon.